File: families/settings/sawtooth_settings/processor/handler.py
# Copyright 2017 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

import logging
import hashlib
import base64
from functools import lru_cache


from sawtooth_sdk.processor.handler import TransactionHandler
from sawtooth_sdk.messaging.future import FutureTimeoutError
from sawtooth_sdk.processor.exceptions import InvalidTransaction
from sawtooth_sdk.processor.exceptions import InternalError

from sawtooth_settings.protobuf.settings_pb2 import SettingsPayload
from sawtooth_settings.protobuf.settings_pb2 import SettingProposal
from sawtooth_settings.protobuf.settings_pb2 import SettingVote
from sawtooth_settings.protobuf.settings_pb2 import SettingCandidate
from sawtooth_settings.protobuf.settings_pb2 import SettingCandidates
from sawtooth_settings.protobuf.setting_pb2 import Setting

LOGGER = logging.getLogger(__name__)


# The config namespace is special: it is not derived from a hash.
SETTINGS_NAMESPACE = '000000'

# Number of seconds to wait for state operations to succeed
STATE_TIMEOUT_SEC = 10


class SettingsTransactionHandler(TransactionHandler):
    @property
    def family_name(self):
        return 'sawtooth_settings'

    @property
    def family_versions(self):
        return ['1.0']

    @property
    def namespaces(self):
        return [SETTINGS_NAMESPACE]

    def apply(self, transaction, context):

        txn_header = transaction.header
        public_key = txn_header.signer_public_key

        auth_keys = _get_auth_keys(context)
        if auth_keys and public_key not in auth_keys:
            raise InvalidTransaction(
                '{} is not authorized to change settings'.format(public_key))

        settings_payload = SettingsPayload()
        settings_payload.ParseFromString(transaction.payload)

        if settings_payload.action == SettingsPayload.PROPOSE:
            return self._apply_proposal(
                auth_keys, public_key, settings_payload.data, context)
        if settings_payload.action == SettingsPayload.VOTE:
            return self._apply_vote(public_key, settings_payload.data,
                                    auth_keys, context)

        raise InvalidTransaction(
            "'action' must be one of {PROPOSE, VOTE} in 'Ballot' mode")

    def _apply_proposal(self, auth_keys, public_key,
                        setting_proposal_data, context):
        setting_proposal = SettingProposal()
        setting_proposal.ParseFromString(setting_proposal_data)

        proposal_id = hashlib.sha256(setting_proposal_data).hexdigest()

        approval_threshold = _get_approval_threshold(context)

        _validate_setting(auth_keys,
                          setting_proposal.setting,
                          setting_proposal.value)

        if approval_threshold > 1:
            setting_candidates = _get_setting_candidates(context)

            existing_candidate = _first(
                setting_candidates.candidates,
                lambda candidate: candidate.proposal_id == proposal_id)

            if existing_candidate is not None:
                raise InvalidTransaction(
                    'Duplicate proposal for {}'.format(
                        setting_proposal.setting))

            record = SettingCandidate.VoteRecord(
                public_key=public_key,
                vote=SettingVote.ACCEPT)
            setting_candidates.candidates.add(
                proposal_id=proposal_id,
                proposal=setting_proposal,
                votes=[record]
            )

            LOGGER.debug('Proposal made to set %s to %s',
                         setting_proposal.setting,
                         setting_proposal.value)
            _save_setting_candidates(context, setting_candidates)
        else:
            _set_setting_value(context,
                               setting_proposal.setting,
                               setting_proposal.value)

    def _apply_vote(self, public_key,
                    settings_vote_data, authorized_keys, context):
        settings_vote = SettingVote()
        settings_vote.ParseFromString(settings_vote_data)
        proposal_id = settings_vote.proposal_id

        setting_candidates = _get_setting_candidates(context)
        candidate = _first(
            setting_candidates.candidates,
            lambda candidate: candidate.proposal_id == proposal_id)

        if candidate is None:
            raise InvalidTransaction(
                "Proposal {} does not exist.".format(proposal_id))

        candidate_index = _index_of(setting_candidates.candidates, candidate)

        approval_threshold = _get_approval_threshold(context)

        vote_record = _first(candidate.votes,
                             lambda record: record.public_key == public_key)
        if vote_record is not None:
            raise InvalidTransaction(
                '{} has already voted'.format(public_key))

        candidate.votes.add(
            public_key=public_key,
            vote=settings_vote.vote)

        accepted_count = 0
        rejected_count = 0
        for vote_record in candidate.votes:
            if vote_record.vote == SettingVote.ACCEPT:
                accepted_count += 1
            elif vote_record.vote == SettingVote.REJECT:
                rejected_count += 1

        if accepted_count >= approval_threshold:
            _set_setting_value(context,
                               candidate.proposal.setting,
                               candidate.proposal.value)
            del setting_candidates.candidates[candidate_index]
        elif rejected_count >= approval_threshold or \
                (rejected_count + accepted_count) == len(authorized_keys):
            LOGGER.debug('Proposal for %s was rejected',
                         candidate.proposal.setting)
            del setting_candidates.candidates[candidate_index]
        else:
            LOGGER.debug('Vote recorded for %s',
                         candidate.proposal.setting)

        _save_setting_candidates(context, setting_candidates)


def _get_setting_candidates(context):
    value = _get_setting_value(context, 'sawtooth.settings.vote.proposals')
    if not value:
        return SettingCandidates(candidates={})

    setting_candidates = SettingCandidates()
    setting_candidates.ParseFromString(base64.b64decode(value))
    return setting_candidates


def _save_setting_candidates(context, setting_candidates):
    _set_setting_value(context,
                       'sawtooth.settings.vote.proposals',
                       base64.b64encode(
                           setting_candidates.SerializeToString()))


def _get_approval_threshold(context):
    return int(_get_setting_value(
        context, 'sawtooth.settings.vote.approval_threshold', 1))


def _get_auth_keys(context):
    value = _get_setting_value(
        context, 'sawtooth.settings.vote.authorized_keys', '')
    return _split_ignore_empties(value)


def _split_ignore_empties(value):
    return [v.strip() for v in value.split(',') if v]


def _validate_setting(auth_keys, setting, value):
    if not auth_keys and \
            setting != 'sawtooth.settings.vote.authorized_keys':
        raise InvalidTransaction(
            'Cannot set {} until authorized_keys is set.'.format(setting))

    if setting == 'sawtooth.settings.vote.authorized_keys':
        if not _split_ignore_empties(value):
            raise InvalidTransaction('authorized_keys must not be empty.')

    if setting == 'sawtooth.settings.vote.approval_threshold':
        threshold = None
        try:
            threshold = int(value)
        except ValueError:
            raise InvalidTransaction('approval_threshold must be an integer')

        if threshold > len(auth_keys):
            raise InvalidTransaction(
                'approval_threshold must be less than or equal to number of '
                'authorized_keys')

    if setting == 'sawtooth.settings.vote.proposals':
        raise InvalidTransaction(
            'Setting sawtooth.settings.vote.proposals is read-only')


def _get_setting_value(context, key, default_value=None):
    address = _make_settings_key(key)
    setting = _get_setting_entry(context, address)
    for entry in setting.entries:
        if key == entry.key:
            return entry.value

    return default_value


def _set_setting_value(context, key, value):
    address = _make_settings_key(key)
    setting = _get_setting_entry(context, address)

    old_value = None
    old_entry_index = None
    for i, entry in enumerate(setting.entries):
        if key == entry.key:
            old_value = entry.value
            old_entry_index = i

    if old_entry_index is not None:
        setting.entries[old_entry_index].value = value
    else:
        setting.entries.add(key=key, value=value)

    try:
        addresses = list(context.set_state(
            {address: setting.SerializeToString()},
            timeout=STATE_TIMEOUT_SEC))
    except FutureTimeoutError:
        LOGGER.warning(
            'Timeout occured on context.set_state([%s, <value>])', address)
        raise InternalError('Unable to set {}'.format(key))

    if len(addresses) != 1:
        LOGGER.warning(
            'Failed to save value on address %s', address)
        raise InternalError(
            'Unable to save config value {}'.format(key))
    if setting != 'sawtooth.settings.vote.proposals':
        LOGGER.info('Setting setting %s changed from %s to %s',
                    key, old_value, value)
    context.add_event(
        event_type="settings/update",
        attributes=[("updated", key)])


def _get_setting_entry(context, address):
    setting = Setting()

    try:
        entries_list = context.get_state([address], timeout=STATE_TIMEOUT_SEC)
    except FutureTimeoutError:
        LOGGER.warning('Timeout occured on context.get_state([%s])', address)
        raise InternalError('Unable to get {}'.format(address))

    if entries_list:
        setting.ParseFromString(entries_list[0].data)

    return setting


def _to_hash(value):
    return hashlib.sha256(value.encode()).hexdigest()


def _first(a_list, pred):
    return next((x for x in a_list if pred(x)), None)


def _index_of(iterable, obj):
    return next((i for i, x in enumerate(iterable) if x == obj), -1)


_MAX_KEY_PARTS = 4
_ADDRESS_PART_SIZE = 16
_EMPTY_PART = _to_hash('')[:_ADDRESS_PART_SIZE]


@lru_cache(maxsize=128)
def _make_settings_key(key):
    # split the key into 4 parts, maximum
    key_parts = key.split('.', maxsplit=_MAX_KEY_PARTS - 1)
    # compute the short hash of each part
    addr_parts = [_to_hash(x)[:_ADDRESS_PART_SIZE] for x in key_parts]
    # pad the parts with the empty hash, if needed
    addr_parts.extend([_EMPTY_PART] * (_MAX_KEY_PARTS - len(addr_parts)))

    return SETTINGS_NAMESPACE + ''.join(addr_parts)

File: families/settings/sawtooth_settings/src/handler.rs
/*
 * Copyright 2018 Bitwise IO, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ------------------------------------------------------------------------------
 */

cfg_if! {
     if #[cfg(target_arch = "wasm32")] {
         use sabre_sdk::ApplyError;
         use sabre_sdk::TransactionContext;
         use sabre_sdk::TransactionHandler;
         use sabre_sdk::TpProcessRequest;
         use sabre_sdk::{WasmPtr, execute_entrypoint};
         use protos::setting::{ Setting,
                        Setting_Entry};
     } else {
         use sawtooth_sdk::messages::processor::TpProcessRequest;
         use sawtooth_sdk::processor::handler::ApplyError;
         use sawtooth_sdk::processor::handler::TransactionContext;
         use sawtooth_sdk::processor::handler::TransactionHandler;
         use sawtooth_sdk::messages::setting::{ Setting,
                                                 Setting_Entry};
    }
}

extern crate base64;
use crypto::digest::Digest;
use crypto::sha2::Sha256;
use protobuf;
use protos::settings::{
    SettingCandidate, SettingCandidate_VoteRecord, SettingCandidates, SettingProposal, SettingVote,
    SettingVote_Vote, SettingsPayload, SettingsPayload_Action,
};
use std::iter::repeat;

#[cfg(target_arch = "wasm32")]
// Sabre apply must return a bool
fn apply(request: &TpProcessRequest, context: &mut TransactionContext) -> Result<bool, ApplyError> {
    let handler = SettingsTransactionHandler::new();
    match handler.apply(request, context) {
        Ok(_) => Ok(true),
        Err(err) => Err(err),
    }
}

#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub unsafe fn entrypoint(payload: WasmPtr, signer: WasmPtr, signature: WasmPtr) -> i32 {
    execute_entrypoint(payload, signer, signature, apply)
}

// The config namespace is special: it is not derived from a hash.
const SETTINGS_NAMESPACE: &str = "000000";

#[derive(Default)]
pub struct SettingsTransactionHandler {
    family_name: String,
    family_versions: Vec<String>,
    namespaces: Vec<String>,
}

impl SettingsTransactionHandler {
    pub fn new() -> SettingsTransactionHandler {
        SettingsTransactionHandler {
            family_name: "sawtooth_settings".to_string(),
            family_versions: vec!["1.0".to_string()],
            namespaces: vec![SETTINGS_NAMESPACE.to_string()],
        }
    }
}

impl TransactionHandler for SettingsTransactionHandler {
    fn family_name(&self) -> String {
        self.family_name.clone()
    }

    fn family_versions(&self) -> Vec<String> {
        self.family_versions.clone()
    }

    fn namespaces(&self) -> Vec<String> {
        self.namespaces.clone()
    }

    fn apply(
        &self,
        transaction: &TpProcessRequest,
        context: &mut dyn TransactionContext,
    ) -> Result<(), ApplyError> {
        let public_key = transaction.get_header().get_signer_public_key();
        let auth_keys = get_auth_keys(context)?;
        if !auth_keys.is_empty() && !auth_keys.contains(&public_key.to_string()) {
            return Err(ApplyError::InvalidTransaction(format!(
                "{:?} is not authorized to change settings",
                public_key
            )));
        }

        let settings_payload: SettingsPayload = unpack_data(transaction.get_payload())?;

        match settings_payload.get_action() {
            SettingsPayload_Action::PROPOSE => apply_proposal(
                &auth_keys,
                &public_key,
                settings_payload.get_data(),
                context,
            ),
            SettingsPayload_Action::VOTE => apply_vote(
                &auth_keys,
                &public_key,
                settings_payload.get_data(),
                context,
            ),
            SettingsPayload_Action::ACTION_UNSET => Err(ApplyError::InvalidTransaction(
                String::from("'action' must be one of {PROPOSE, VOTE} in 'Ballot' mode"),
            )),
        }
    }
}

fn apply_proposal(
    auth_keys: &[String],
    public_key: &str,
    setting_proposal_data: &[u8],
    context: &mut dyn TransactionContext,
) -> Result<(), ApplyError> {
    let setting_proposal: SettingProposal = unpack_data(setting_proposal_data)?;

    let proposal_id = proposal_to_hash(setting_proposal_data);

    let approval_threshold = get_approval_threshold(context)?;

    validate_setting(
        auth_keys,
        setting_proposal.get_setting(),
        setting_proposal.get_value(),
    )?;

    if approval_threshold > 1 {
        let mut setting_candidates = get_setting_candidates(context)?;
        for candidate in setting_candidates.get_candidates().iter() {
            if candidate.get_proposal_id() == proposal_id {
                return Err(ApplyError::InvalidTransaction(format!(
                    "Duplicate proposal for {:?}",
                    setting_proposal.get_setting()
                )));
            }
        }

        let mut vote_record = SettingCandidate_VoteRecord::new();
        vote_record.set_public_key(public_key.to_string());
        vote_record.set_vote(SettingVote_Vote::ACCEPT);

        let mut candidate = SettingCandidate::new();
        candidate.set_proposal_id(proposal_id);
        candidate.set_proposal(setting_proposal);
        candidate.set_votes(protobuf::RepeatedField::from_vec(vec![vote_record]));
        setting_candidates.mut_candidates().push(candidate);

        save_settings_candidates(context, &setting_candidates)
    } else {
        set_setting_value(
            context,
            setting_proposal.get_setting(),
            setting_proposal.get_value(),
        )
    }
}

fn apply_vote(
    auth_keys: &[String],
    public_key: &str,
    setting_vote_data: &[u8],
    context: &mut dyn TransactionContext,
) -> Result<(), ApplyError> {
    let setting_vote: SettingVote = unpack_data(setting_vote_data)?;
    let proposal_id = setting_vote.get_proposal_id();

    let mut setting_candidates = get_setting_candidates(context)?;

    let mut accepted_count = 0;
    let mut rejected_count = 0;
    let candidate_value;
    let candidate_setting;
    let candidate_index;

    {
        candidate_index = get_candidate_index(&mut setting_candidates, proposal_id)?;
        let candidate = setting_candidates
            .mut_candidates()
            .get_mut(candidate_index)
            .unwrap();

        for record in candidate.get_votes().iter() {
            if record.get_public_key() == public_key {
                return Err(ApplyError::InvalidTransaction(format!(
                    "{:?} already has voted",
                    public_key
                )));
            }
        }

        let mut vote_record = SettingCandidate_VoteRecord::new();
        vote_record.set_public_key(public_key.to_string());
        vote_record.set_vote(setting_vote.get_vote());
        candidate.mut_votes().push(vote_record);

        for record in candidate.get_votes().iter() {
            match record.get_vote() {
                SettingVote_Vote::ACCEPT => accepted_count += 1,
                SettingVote_Vote::REJECT => rejected_count += 1,
                _ => {
                    return Err(ApplyError::InvalidTransaction(String::from(
                        "Transaction type unset",
                    )));
                }
            }
        }

        candidate_value = candidate.get_proposal().get_value().to_string();
        candidate_setting = candidate.get_proposal().get_setting().to_string();
    }

    let approval_threshold = get_approval_threshold(context)?;
    if accepted_count >= approval_threshold {
        set_setting_value(context, &candidate_setting, &candidate_value)?;
        setting_candidates.mut_candidates().remove(candidate_index);
    } else if rejected_count >= approval_threshold
        || rejected_count + accepted_count == auth_keys.len() as i32
    {
        debug!("Proposal for {} was rejected", &candidate_setting);
        setting_candidates.mut_candidates().remove(candidate_index);
    } else {
        debug!("Vote recorded for {}", &candidate_setting);
    }
    save_settings_candidates(context, &setting_candidates)
}

fn unpack_data<T>(data: &[u8]) -> Result<T, ApplyError>
where
    T: protobuf::Message,
{
    protobuf::parse_from_bytes(&data).map_err(|err| {
        warn!(
            "Invalid error: Failed to unmarshal SettingsTransaction: {:?}",
            err
        );
        ApplyError::InternalError(format!(
            "Failed to unmarshal SettingsTransaction: {:?}",
            err
        ))
    })
}

fn get_candidate_index(
    setting_candidates: &mut SettingCandidates,
    proposal_id: &str,
) -> Result<usize, ApplyError> {
    for (i, setting_candidate) in setting_candidates.mut_candidates().iter_mut().enumerate() {
        if setting_candidate.get_proposal_id() == proposal_id {
            return Ok(i);
        }
    }
    Err(ApplyError::InvalidTransaction(format!(
        "Proposal {:?} does not exist",
        proposal_id
    )))
}

fn save_settings_candidates(
    context: &mut dyn TransactionContext,
    setting_candidates: &SettingCandidates,
) -> Result<(), ApplyError> {
    let data = protobuf::Message::write_to_bytes(setting_candidates).map_err(|err| {
        ApplyError::InternalError(format!("Failed to serialize SettingsCandidates: {:?}", err))
    })?;
    set_setting_value(
        context,
        "sawtooth.settings.vote.proposals",
        &base64::encode(&data),
    )
}

fn set_setting_value(
    context: &mut dyn TransactionContext,
    key: &str,
    value: &str,
) -> Result<(), ApplyError> {
    let address = make_settings_key(key);
    let setting_data = get_setting_data(&address, context)?;
    let mut setting = match setting_data {
        None => Setting::new(),
        Some(entries) => unpack_data(&entries)?,
    };

    let old_index = get_entry_index(&setting, key);

    match old_index {
        None => {
            let mut setting_entry = Setting_Entry::new();
            setting_entry.set_key(key.to_string());
            setting_entry.set_value(value.to_string());
            setting.mut_entries().push(setting_entry);
        }
        Some(i) => {
            setting.mut_entries()[i].set_value(value.to_string());
        }
    }

    set_state(context, key, value, &address, &setting)
}

fn set_state(
    context: &mut dyn TransactionContext,
    key: &str,
    value: &str,
    address: &str,
    setting: &Setting,
) -> Result<(), ApplyError> {
    let data = protobuf::Message::write_to_bytes(setting).map_err(|err| {
        ApplyError::InternalError(format!("Failed to serialize Setting: {:?}", err))
    })?;

    context
        .set_state_entry(address.to_string(), data)
        .map_err(|_| {
            warn!("Failed to save value on address {}", &address);
            ApplyError::InternalError(format!("Unable to save config value {}", key))
        })?;
    if key != "sawtooth.settings.vote.proposals" {
        info!("Setting {:?} changed to {:?}", key, value);
    }
    #[cfg(not(target_arch = "wasm32"))]
    context
        .add_event(
            "settings/update".to_string(),
            vec![("updated".to_string(), key.to_string())],
            &[],
        )
        .map_err(|_| {
            warn!("Failed to add event {}", key);
            ApplyError::InternalError(format!("Failed to add event {}", key))
        })?;
    Ok(())
}

fn get_entry_index(setting: &Setting, key: &str) -> Option<usize> {
    for (i, entry) in setting.get_entries().iter().enumerate() {
        if entry.get_key() == key {
            return Some(i);
        }
    }
    None
}

fn get_setting_candidates(
    context: &mut dyn TransactionContext,
) -> Result<SettingCandidates, ApplyError> {
    let value = get_setting_value(context, "sawtooth.settings.vote.proposals")?;
    match value {
        None => Ok(SettingCandidates::new()),
        Some(val) => {
            let v = base64::decode(&val).map_err(|err| {
                ApplyError::InternalError(format!(
                    "Cannot decode sawtooth.settings.vote.proposals: {:?} ",
                    err
                ))
            })?;
            unpack_data(&v)
        }
    }
}

fn validate_setting(auth_keys: &[String], setting: &str, value: &str) -> Result<(), ApplyError> {
    if auth_keys.is_empty() && setting != "sawtooth.settings.vote.authorized_keys" {
        return Err(ApplyError::InvalidTransaction(format!(
            "Cannot set {:?} until authorized_keys is set.",
            setting
        )));
    }
    if setting == "sawtooth.settings.vote.authorized_keys" && split_ignore_empties(value).is_empty()
    {
        return Err(ApplyError::InvalidTransaction(
            "authorized_keys must not be empty.".into(),
        ));
    }
    if setting == "sawtooth.settings.vote.approval_threshold" {
        let threshold = value.parse::<i32>().map_err(|_| {
            warn!("Failed to parse sawtooth.settings.vote.approval_threshold value");
            ApplyError::InvalidTransaction("approval_threshold must be an integer".into())
        })?;
        if threshold > auth_keys.len() as i32 {
            return Err(ApplyError::InvalidTransaction(
                "approval_threshold must be less than or equal to number of authorized_keys".into(),
            ));
        }
    }
    if setting == "sawtooth.settings.vote.proposals" {
        return Err(ApplyError::InvalidTransaction(
            "Setting sawtooth.settings.vote.proposals is read-only".into(),
        ));
    }
    Ok(())
}

fn get_auth_keys(context: &mut dyn TransactionContext) -> Result<Vec<String>, ApplyError> {
    let auth_keys = get_setting_value(context, "sawtooth.settings.vote.authorized_keys")?;
    match auth_keys {
        None => Ok(vec![]),
        Some(value) => Ok(split_ignore_empties(&value)),
    }
}
fn split_ignore_empties(values: &str) -> Vec<String> {
    values
        .split(',')
        .map(|v| v.trim().to_string())
        .filter(|v| !v.is_empty())
        .collect()
}

fn proposal_to_hash(value: &[u8]) -> String {
    let mut sha = Sha256::new();
    sha.input(value);
    sha.result_str()
}

fn get_approval_threshold(context: &mut dyn TransactionContext) -> Result<i32, ApplyError> {
    let settings_value = get_setting_value(context, "sawtooth.settings.vote.approval_threshold")?;
    match settings_value {
        None => Ok(1), //1 is the default_value for approval_threshold
        Some(value) => {
            let val = value.parse::<i32>().map_err(|err| {
                warn!("Failed to parse sawtooth.settings.vote.approval_threshold value");
                ApplyError::InternalError(format!("Failed to parse
                                              sawtooth.settings.vote.approval_threshold value: {:?}", err))
            })?;
            Ok(val)
        }
    }
}

fn get_setting_value(
    context: &mut dyn TransactionContext,
    key: &str,
) -> Result<Option<String>, ApplyError> {
    let address = make_settings_key(key);
    let setting_data = get_setting_data(&address, context)?;

    match setting_data {
        None => Ok(None),
        Some(entries) => {
            let setting: Setting = unpack_data(&entries)?;
            for entry in setting.get_entries().iter() {
                if entry.get_key() == key {
                    return Ok(Some(entry.get_value().to_string()));
                }
            }
            Ok(None)
        }
    }
}

fn get_setting_data(
    address: &str,
    context: &mut dyn TransactionContext,
) -> Result<Option<Vec<u8>>, ApplyError> {
    context.get_state_entry(address).map_err(|err| {
        warn!("Internal Error: Failed to load state: {:?}", err);
        ApplyError::InternalError(format!("Failed to load state: {:?}", err))
    })
}

const MAX_KEY_PARTS: usize = 4;
const ADDRESS_PART_SIZE: usize = 16;

fn make_settings_key(key: &str) -> String {
    let mut address = String::new();
    address.push_str(SETTINGS_NAMESPACE);
    address.push_str(
        &key.splitn(MAX_KEY_PARTS, '.')
            .chain(repeat(""))
            .map(short_hash)
            .take(MAX_KEY_PARTS)
            .collect::<Vec<_>>()
            .join(""),
    );

    address
}

fn short_hash(s: &str) -> String {
    let mut sha = Sha256::new();
    sha.input(s.as_bytes());
    sha.result_str()[..ADDRESS_PART_SIZE].to_string()
}
